source/extensions/quic_listeners/quiche/envoy_quic_server_connection.cc

```diff
@@ -0,0 +1,60 @@
+#include "extensions/quic_listeners/quiche/envoy_quic_server_connection.h"
+
+#include "common/network/listen_socket_impl.h"
+
+#include "extensions/quic_listeners/quiche/envoy_quic_utils.h"
+#include "extensions/quic_listeners/quiche/quic_io_handle_wrapper.h"
+#include "extensions/transport_sockets/well_known_names.h"
+
+namespace Envoy {
+namespace Quic {
+
+EnvoyQuicServerConnection::EnvoyQuicServerConnection(
+    const quic::QuicConnectionId& server_connection_id,
+    quic::QuicSocketAddress initial_peer_address, quic::QuicConnectionHelperInterface& helper,
+    quic::QuicAlarmFactory& alarm_factory, quic::QuicPacketWriter* writer, bool owns_writer,
+    const quic::ParsedQuicVersionVector& supported_versions,
+    Network::ListenerConfig& listener_config, Server::ListenerStats& listener_stats)
+    : EnvoyQuicConnection(
+          server_connection_id, initial_peer_address, helper, alarm_factory, writer, owns_writer,
+          quic::Perspective::IS_SERVER, supported_versions,
+          std::make_unique<Network::ConnectionSocketImpl>(
+              // Wraps the real IoHandle instance so that if the connection socket gets closed,
+              // the real IoHandle won't be affected.
+              std::make_unique<QuicIoHandleWrapper>(listener_config.socket().ioHandle()), nullptr,
+              quicAddressToEnvoyAddressInstance(initial_peer_address))),
+      listener_config_(listener_config), listener_stats_(listener_stats) {}
+
+bool EnvoyQuicServerConnection::OnPacketHeader(const quic::QuicPacketHeader& header) {
+  if (!EnvoyQuicConnection::OnPacketHeader(header)) {
+    return false;
+  }
+  if (connectionSocket()->localAddress() != nullptr) {
+    return true;
+  }
+  ASSERT(self_address().IsInitialized());
+  // Self address should be initialized by now. It's time to install filters.
+  connectionSocket()->setLocalAddress(quicAddressToEnvoyAddressInstance(self_address()));
+  connectionSocket()->setDetectedTransportProtocol(
+      Extensions::TransportSockets::TransportSocketNames::get().Quic);
+  ASSERT(filter_chain_ == nullptr);
+  filter_chain_ = listener_config_.filterChainManager().findFilterChain(*connectionSocket());
```

Comment:
Can there ever be a case in which we don't have a filter chain and a non-empty one? Isn't all of this hard coded and/or checked by config?

Reply:
There should be at least one filter chain for HCM. It's configurable via config, not hard coded. And HCM setup relies on filter manager to read from config as a standard filter.  Please see this example of config to setup QUIC listener and HCM:  https://github.com/envoyproxy/envoy/compare/master...danzh2010:quicclient?expand=1#diff-6e3a243807199766bdae36c0c2f242eb in test/config/utility.cc